From 6ef9ac191fcd170bb0f9417e5f5b6c590163fd6f Mon Sep 17 00:00:00 2001
From: Shammah Chancellor <shammah.chancellor@gmail.com>
Date: Sun, 16 May 2021 09:38:01 -0700
Subject: [PATCH 099/196] Implement raw encoding and decoding on XAddress
 format (#172)

This is the first commit required to implement parsing XAddress
formats. The basic idea is that addresses are composed of 4 parts:

1. token prefix (all lowercase word characters)
2. network byte (capital word character)
3. base58 encoded payload
4. base58 encoded 4-byte weak sha256d hash of (1)|| (2) || (3)

A follow up commit will implement processing into a `CTxDestination`
to support use in the rest of the node.
---
 src/CMakeLists.txt          |  1 +
 src/addresses/xaddress.cpp  | 84 +++++++++++++++++++++++++++++++
 src/addresses/xaddress.h    | 59 ++++++++++++++++++++++
 src/test/CMakeLists.txt     |  1 +
 src/test/xaddress_tests.cpp | 98 +++++++++++++++++++++++++++++++++++++
 5 files changed, 243 insertions(+)
 create mode 100644 src/addresses/xaddress.cpp
 create mode 100644 src/addresses/xaddress.h
 create mode 100644 src/test/xaddress_tests.cpp

diff --git a/src/CMakeLists.txt b/src/CMakeLists.txt
index 970d2a56f..057ff1867 100644
--- a/src/CMakeLists.txt
+++ b/src/CMakeLists.txt
@@ -459,6 +459,7 @@ add_library(common
 	bloom.cpp
 	cashaddr.cpp
 	cashaddrenc.cpp
+	addresses/xaddress.cpp
 	chainparams.cpp
 	config.cpp
 	consensus/merkle.cpp
diff --git a/src/addresses/xaddress.cpp b/src/addresses/xaddress.cpp
new file mode 100644
index 000000000..df7bcf25c
--- /dev/null
+++ b/src/addresses/xaddress.cpp
@@ -0,0 +1,84 @@
+// Copyright (c) 2021 Logos Foundation
+// Distributed under the MIT software license, see the accompanying
+// file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+#include <addresses/xaddress.h>
+#include <base58.h>
+#include <hash.h>
+#include <uint256.h>
+#include <util/strencodings.h>
+#include <util/string.h>
+#include <util/vector.h>
+
+#include <string>
+
+namespace XAddress {
+/**
+ * Create SHA256 Hash of the address contents for integrity checks.
+ */
+uint256 HashAddressContents(const Content &addressContent) {
+    CHashWriter hasher(SER_GETHASH, 0);
+    hasher << addressContent.m_token;
+    hasher << uint8_t(addressContent.m_network);
+    hasher << uint8_t(addressContent.m_type);
+    hasher << addressContent.m_payload;
+    return hasher.GetSHA256();
+}
+
+/**
+ * Encode a XAddress string.
+ */
+std::string Encode(const Content &addressContent) {
+    std::vector<uint8_t> preencodedBuffer;
+    uint256 check = HashAddressContents(addressContent);
+    preencodedBuffer.reserve(addressContent.m_payload.size() + 5);
+    preencodedBuffer.push_back(addressContent.m_type);
+    preencodedBuffer.insert(preencodedBuffer.end(),
+                            addressContent.m_payload.begin(),
+                            addressContent.m_payload.end());
+    preencodedBuffer.insert(preencodedBuffer.end(), check.begin(),
+                            check.begin() + 4);
+    std::string address;
+    address.append(addressContent.m_token);
+    address.append(1, addressContent.m_network);
+    address.append(EncodeBase58(preencodedBuffer));
+    return address;
+}
+
+/**
+ * Decode a XAddress string.
+ */
+DecodeError Decode(const std::string &address, Content &parsedOutput) {
+    const auto networkPosition =
+        std::find_if(address.begin(), address.end(), [](const char c) -> bool {
+            return std::isupper(c) || std::isdigit(c) || c == '_';
+        });
+    if (networkPosition == address.end()) {
+        return NO_NETWORK_POSITION;
+    }
+    const std::string token =
+        address.substr(0, std::distance(address.begin(), networkPosition));
+    const NetworkType networkByte = NetworkType(*networkPosition);
+
+    // Need to be able to get a c_str() from this, so we must do a copy.
+    const std::string encodedPayload(networkPosition + 1, address.end());
+    // vch length cannot be greater than encodedPayload
+    std::vector<uint8_t> vch;
+    if (!DecodeBase58(encodedPayload, vch, encodedPayload.length())) {
+        return BASE58_DECODE_FAILED;
+    }
+    // Undersized payload. Can't include type byte and checksum
+    if (vch.size() < 5) {
+        return UNDERSIZED_PAYLOAD;
+    }
+    const AddressType addressByte = AddressType(vch[0]);
+    parsedOutput = Content(token, networkByte, addressByte,
+                           std::vector(vch.begin() + 1, vch.end() - 4));
+    const uint256 check = HashAddressContents(parsedOutput);
+    if (memcmp(&check, &vch[vch.size() - 4], 4)) {
+        return INTEGRITY_CHECK_FAILED;
+    }
+    return DECODE_OK;
+}
+
+} // namespace XAddress
diff --git a/src/addresses/xaddress.h b/src/addresses/xaddress.h
new file mode 100644
index 000000000..1e0f6055a
--- /dev/null
+++ b/src/addresses/xaddress.h
@@ -0,0 +1,59 @@
+// Copyright (c) 2021 Logos Foundation
+// Distributed under the MIT software license, see the accompanying
+// file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+#ifndef BITCOIN_XADDR_H
+#define BITCOIN_XADDR_H
+
+#include <string>
+#include <vector>
+
+namespace XAddress {
+
+const std::string TOKEN_NAME = "lotus";
+
+enum NetworkType : uint8_t {
+    MAINNET = '_',
+    TESTNET = 'T',
+    REGTEST = 'R',
+};
+
+enum AddressType : uint8_t {
+    SCRIPT_PUB_KEY = 0,
+};
+
+enum DecodeError : uint8_t {
+    DECODE_OK = 0,
+    NO_NETWORK_POSITION = 1,
+    BASE58_DECODE_FAILED = 2,
+    UNDERSIZED_PAYLOAD = 3,
+    INTEGRITY_CHECK_FAILED = 4,
+};
+
+class Content {
+public:
+    std::string m_token;
+    NetworkType m_network;
+    AddressType m_type;
+    std::vector<uint8_t> m_payload;
+
+    Content() = default;
+
+    Content(std::string token, NetworkType network, AddressType type,
+            std::vector<uint8_t> payload)
+        : m_token(token), m_network(network), m_type(type), m_payload(payload) {
+    }
+};
+
+/**
+ * Encode an XAddress string.
+ */
+std::string Encode(const Content &addressContent);
+/**
+ * Decode an XAddress string. Returns false on failure, and parsedOutput is not
+ * modified.
+ */
+DecodeError Decode(const std::string &address, Content &parsedOutput);
+} // namespace XAddress
+
+#endif // BITCOIN_XADDR_H
diff --git a/src/test/CMakeLists.txt b/src/test/CMakeLists.txt
index 0e5e7b4fa..a3d87de2e 100644
--- a/src/test/CMakeLists.txt
+++ b/src/test/CMakeLists.txt
@@ -138,6 +138,7 @@ add_boost_unit_tests_to_suite(bitcoin test_lotus
 		bswap_tests.cpp
 		cashaddr_tests.cpp
 		cashaddrenc_tests.cpp
+		xaddress_tests.cpp
 		checkdatasig_tests.cpp
 		checkpoints_tests.cpp
 		checkqueue_tests.cpp
diff --git a/src/test/xaddress_tests.cpp b/src/test/xaddress_tests.cpp
new file mode 100644
index 000000000..4599acaea
--- /dev/null
+++ b/src/test/xaddress_tests.cpp
@@ -0,0 +1,98 @@
+// Copyright (c) 2021 Logos Foundation
+// Distributed under the MIT software license, see the accompanying
+// file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+#include <addresses/xaddress.h>
+#include <random.h>
+#include <test/util/setup_common.h>
+#include <util/strencodings.h>
+#include <util/vector.h>
+
+#include <test/lcg.h>
+
+#include <algorithm>
+#include <boost/test/unit_test.hpp>
+#include <functional>
+#include <random>
+#include <string>
+
+/** All alphanumeric characters except for "0", "I", "O", and "l" */
+const std::string pszBase58 =
+    "123456789ABCDEFGHJKLMNPQRSTUVWXYZabcdefghijkmnopqrstuvwxyz";
+
+BOOST_FIXTURE_TEST_SUITE(xaddress_tests, BasicTestingSetup)
+
+BOOST_AUTO_TEST_CASE(raw_encode_decode_succeeds) {
+    std::random_device rd;
+    auto gen = [&rd]() { return uint8_t(rd()); };
+
+    // Assert that encode/decode works.
+    for (size_t i = 0; i < 10000; i++) {
+        std::vector<uint8_t> payload(1);
+        std::generate(begin(payload), end(payload), gen);
+        const std::string encodedAddress = XAddress::Encode(
+            XAddress::Content(XAddress::TOKEN_NAME, XAddress::MAINNET,
+                              XAddress::SCRIPT_PUB_KEY, payload));
+        XAddress::Content parsedAddress;
+        BOOST_CHECK_EQUAL(XAddress::Decode(encodedAddress, parsedAddress),
+                          XAddress::DECODE_OK);
+        BOOST_CHECK_EQUAL(parsedAddress.m_token, XAddress::TOKEN_NAME);
+        BOOST_CHECK_MESSAGE(parsedAddress.m_network == XAddress::MAINNET,
+                            "Network byte incorrect");
+        BOOST_CHECK_EQUAL(parsedAddress.m_type, XAddress::SCRIPT_PUB_KEY);
+        BOOST_CHECK_MESSAGE(parsedAddress.m_payload == payload,
+                            "Parsed payload incorrect");
+    }
+}
+
+BOOST_AUTO_TEST_CASE(raw_encode_decode_fails) {
+    std::random_device rd;
+    auto gen = [&rd]() { return uint8_t(rd()); };
+
+    // Assert that we get some XAddress::BASE58_DECODE_FAILED or
+    // XAddress::INTEGRITY_CHECK_FAILED failures.
+    for (size_t i = 0; i < 1000; i++) {
+        std::vector<uint8_t> payload(20);
+        std::generate(begin(payload), end(payload), gen);
+        std::string encodedAddress = XAddress::Encode(
+            XAddress::Content(XAddress::TOKEN_NAME, XAddress::MAINNET,
+                              XAddress::SCRIPT_PUB_KEY, payload));
+        const size_t mutatedPosition = rd() % encodedAddress.size();
+        const unsigned char replacedCharacter = rd() % pszBase58.size();
+        const auto error = encodedAddress[mutatedPosition] == replacedCharacter
+                               ? XAddress::DECODE_OK
+                               : XAddress::BASE58_DECODE_FAILED |
+                                     XAddress::INTEGRITY_CHECK_FAILED;
+        encodedAddress[mutatedPosition] = replacedCharacter;
+        XAddress::Content parsedAddress;
+        const auto decodeResult =
+            XAddress::Decode(encodedAddress, parsedAddress);
+        // Error of one of the 2 types or passes
+        BOOST_CHECK_MESSAGE((decodeResult & ~error) == 0,
+                            "Decode or check failed: " << int(decodeResult));
+    }
+
+    // Assert that we receive a NO_NETWORK_POSITION error.
+    {
+        XAddress::Content parsedAddress;
+        const auto decodeResult =
+            XAddress::Decode("lotusabcfdsafdsafdsfdasfdsa", parsedAddress);
+        // Error of one of the 2 types or passes
+        BOOST_CHECK_MESSAGE(decodeResult == XAddress::NO_NETWORK_POSITION,
+                            "Should have errored as NO_NETWORK_POSITION: "
+                                << int(decodeResult));
+    }
+
+    // Assert that undersized payloads fail.
+    {
+        XAddress::Content parsedAddress;
+        const auto decodeResult =
+            XAddress::Decode("lotus_fdasf", parsedAddress);
+        // Error of one of the 2 types or passes
+        BOOST_CHECK_MESSAGE(decodeResult == XAddress::UNDERSIZED_PAYLOAD,
+                            "Should have errored as NO_NETWORK_POSITION: "
+                                << int(decodeResult));
+    }
+}
+
+BOOST_AUTO_TEST_SUITE_END()
-- 
2.39.5 (Apple Git-154)

